mixer/adapter/cloudwatch/logHandler_test.go

```diff
@@ -0,0 +1,337 @@
+	template := `{{or (.sourceIp) "-"}} - {{or (.sourceUser) "-"}}`
+	cfg := &config.Params{
+		LogGroupName:  "testLogGroup",
+		LogStreamName: "testLogStream",
+		Logs: map[string]*config.Params_LogInfo{
+			"accesslog": {
+				PayloadTemplate: template,
+			},
+		},
+	}
+	emptyPayloadCfg := &config.Params{
+		LogGroupName:  "testLogGroup",
+		LogStreamName: "testLogStream",
+		Logs: map[string]*config.Params_LogInfo{
+			"accesslog": {
+				PayloadTemplate: "",
+			},
+		},
+	}
+
+	cases := []struct {
+		handler              adapter.Handler
+		insts                []*logentry.Instance
+		expectedLogEntryData []*cloudwatchlogs.InputLogEvent
+	}{
+		// empty instances
+		{
+			newHandler(nil, nil, env, cfg, &mockCloudWatchClient{}, &mockCloudWatchLogsClient{resp: generateLogStreamOutput(), describeError: nil}),
+			[]*logentry.Instance{},
+			[]*cloudwatchlogs.InputLogEvent{},
+		},
+		// non empty instances
+		{
+			newHandler(nil, nil, env, cfg, &mockCloudWatchClient{}, &mockCloudWatchLogsClient{resp: generateLogStreamOutput(), describeError: nil}),
+			[]*logentry.Instance{
+				{
+					Timestamp: timestp,
+					Name:      "accesslog",
+					Severity:  "Default",
+					Variables: map[string]interface{}{
+						"sourceUser": "abc",
+					},
+				},
+			},
+			[]*cloudwatchlogs.InputLogEvent{
+				{
+					Message:   aws.String("- - abc"),
+					Timestamp: aws.Int64(timestp.UnixNano() / int64(time.Millisecond)),
+				},
+			},
+		},
+		{
+			newHandler(nil, nil, env, cfg, &mockCloudWatchClient{}, &mockCloudWatchLogsClient{resp: generateLogStreamOutput(), describeError: nil}),
+			[]*logentry.Instance{
+				{
+					Timestamp: timestp,
+					Name:      "accesslog",
+					Severity:  "Default",
+					Variables: map[string]interface{}{
+						"sourceUser": "abc",
+						"sourceIp":   "10.0.0.0",
+					},
+				},
+			},
+			[]*cloudwatchlogs.InputLogEvent{
+				{
+					Message:   aws.String("10.0.0.0 - abc"),
+					Timestamp: aws.Int64(timestp.UnixNano() / int64(time.Millisecond)),
+				},
+			},
+		},
+		// payload template not provided explicitly
+		{
+			newHandler(nil, nil, env, emptyPayloadCfg, &mockCloudWatchClient{}, &mockCloudWatchLogsClient{resp: generateLogStreamOutput(), describeError: nil}),
+			[]*logentry.Instance{
+				{
+					Timestamp: timestp,
+					Name:      "accesslog",
+					Severity:  "Default",
+					Variables: map[string]interface{}{
+						"sourceUser": "abc",
+						"sourceIp":   "10.0.0.0",
+					},
+				},
+			},
+			[]*cloudwatchlogs.InputLogEvent{
+				{
+					Message:   aws.String(""),
+					Timestamp: aws.Int64(timestp.UnixNano() / int64(time.Millisecond)),
+				},
+			},
+		},
+	}
+
+	for _, c := range cases {
+		h, ok := c.handler.(*handler)
+		if !ok {
+			t.Error("test case has the wrong type of handler.")
+		}
+
+		ld := h.generateLogEntryData(c.insts)
+
+		if len(c.expectedLogEntryData) != len(ld) {
+			t.Errorf("expected %v logentry data items but got %v", len(c.expectedLogEntryData), len(ld))
```

Comment:
nit: here and below, let's use the golang convention of got, then want.

here, I'd use something like `t.Errorf("generateLogEntryData() generated %d items; wanted %d", ...)`

> Note that the order here is actual != expected, and the message uses that order too. Some test frameworks encourage writing these backwards: 0 != x, "expected 0, got x", and so on. Go does not.

https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures

Reply:
got, then want.
done.